File: python/52. N-Queens II.py
class Solution:
    def totalNQueens(self, n):
        """
        :type n: int
        :rtype: int
        """
        S = [0] * (n+1)        # 行安全
        U = [0] * (2*n + 1)    # 记录col + row，即向上的对角线是否安全；从左下到右上:1、2...2*n
        D = [0] * (2*n + 1)    # 记录col - row + (n+1), 即向下的对角是否安全；从左下到右上：2*n、2*n-1...3、2、1
        cnt = 0
        return self.findPosition(cnt, 1, n, S, U, D)

    def findPosition(self, cnt, col, n, S, U, D):
        if col == n + 1:
            cnt += 1
            return cnt
        for row in range(1, n+1):
            if (S[row] == 0 and U[col+row] == 0 and D[col - row + (n + 1)] == 0):
                S[row] = 1
                U[col + row] = 1
                D[col - row + (n + 1)] = 1
                cnt = self.findPosition(cnt, col+1, n, S, U, D)
                S[row] = 0
                U[col + row] = 0
                D[col - row + (n + 1)] = 0
        return cnt


if __name__ == '__main__':
    print(Solution().totalNQueens(4))



 

        

File: cpp/52.N-Queens II.cpp
# include <iostream>
using namespace std;

class Solution {
public:
    int totalNQueens(int n) {
        int R[n+1];
		int U[2*n + 1];
		int D[2*n + 1];
		for (int i=0; i<n+1; i++)
			R[i] = 0;
		for (int i=0; i<2*n+1; i++){
			U[i] = 0;
			D[i] = 0;
 		}
		int cnt = 0; 
		return getCounts(cnt, 1, R, U, D, n);
    }
private:
	int getCounts(int cnt, int col, int R[], int U[], int D[], int n){
		if (col == n+1){
			++cnt;
			return cnt;
		}
		for (int row = 1; row <= n; row++){
			if (R[row]==0 && U[col+row]==0 and D[col-row+(n+1)]==0){
				R[row]=1;
				U[col+row]=1;
				D[col-row+(n+1)]=1;
				cnt = getCounts(cnt, col+1, R, U, D, n);
				R[row]=0;
				U[col+row]=0;
				D[col-row+(n+1)]=0;
			}
		}
		return cnt;
	}
};
